torchx/lib/torchx/backend.ex

```diff
@@ -125,11 +125,56 @@ defmodule Torchx.Backend do
 
   ## Transfer
 
-  # TODO: Handle backend_options
   @impl true
-  def to_batched_list(%T{shape: shape} = out, %T{} = t, _backend_options) do
-    Torchx.split(from_nx(t), elem(shape, 0))
-    |> Enum.map(&to_nx(&1, out))
+  def to_batched_list(%T{shape: shape} = out, %T{} = t, opts) do
+    leftover = opts[:leftover]
+
+    batch_size = elem(shape, 0)
+    t_axis_0 = elem(t.shape, 0)
+
+    remainder = rem(t_axis_0, batch_size)
+    num_full_batches = div(t_axis_0, batch_size)
+
+    num_batches =
+      if leftover == :repeat and remainder != 0 do
+        num_full_batches + 1
+      else
+        num_full_batches
+      end
+
+    to_batch =
+      cond do
+        remainder != 0 and leftover == :discard ->
+          t
+
+        remainder != 0 and leftover == :repeat ->
+          concat_shape =
+            t.shape |> Tuple.delete_at(0) |> Tuple.insert_at(0, num_batches * batch_size)
+
+          slice_length = t_axis_0 - remainder
+          slice_shape = t.shape |> Tuple.delete_at(0) |> Tuple.insert_at(0, slice_length)
+
+          slice = slice(%{t | shape: slice_shape}, t, [0], [slice_length], [1])
+
+          concatenate(%{t | shape: concat_shape}, [t, slice], 0)
+
+        true ->
+          t
+      end
+
+    batched =
+      to_batch
+      |> from_nx()
+      |> Torchx.split(batch_size)
+      |> Enum.take(num_batches)
```

Comment:
@josevalim I'm not sure what's happening, but the `slice` and `concatenate` above are effectively concatenating two full copies of the tensor. This `Enum.take` ends up being necessary because otherwise I get more batches than needed. Is this "soft slicing" expected?

This seems to be a behavior with slice itself, which can be checked by manually extending the shape of the resulting tensor.

Reply:
```elixir
iex(9)> t = Nx.tensor([0, 0, 0, 1, 1, 1])
#Nx.Tensor<
  s64[6]
  Torchx.Backend(cpu)
  [0, 0, 0, 1, 1, 1]
>
iex(10)> s = Nx.slice(t, [0], [3])
#Nx.Tensor<
  s64[3]
  Torchx.Backend(cpu)
  [0, 0, 0]
>
iex(11)> Nx.concatenate([t, s, t])
#Nx.Tensor<
  s64[15]
  Torchx.Backend(cpu)
  [0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0]
>
iex(12)> Nx.concatenate([t, t, s])
#Nx.Tensor<
  s64[15]
  Torchx.Backend(cpu)
  [0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0]
>
```

The underlying issue is that `slice` is not actually getting rid of the underlying data of the tensor, and this is somehow leaking through `concatenate`, but can also be observed with `Torchx.to_blob/1`:

```elixir
iex(14)> Torchx.to_blob(Torchx.from_nx(s))
<<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
  0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0>>
iex(15)> s
#Nx.Tensor<
  s64[3]
  Torchx.Backend(cpu)
  [0, 0, 0]
>
```

Reply:
Maybe slice is returning a view only? And then when you concatenate it goes against the actual tensor?

Reply:
Yeah, I think that's the case. Even if I apply narrow_copy or .detach().clone().resize_(..) this same behavior persists, so probably there's another function for this use case.

Reply:
In the examples I found, narrow was always used for in-place updates

Reply:
Do we still need Enum.take after the slice fix?